File: basic.py
# Cargo Manifest Validator - Python Implementation

def normalizeUnits(manifest):
    """
    Normalizes the weight to kilograms without mutating the original manifest
    
    Args:
        manifest (dict): The cargo manifest object
        
    Returns:
        dict: A new manifest object with weight in kg
    """
    # Create a copy of the manifest to avoid mutation
    normalized = manifest.copy()
    
    # Convert pounds to kilograms if necessary
    if normalized.get("unit") == "lb":
        normalized["weight"] = normalized["weight"] * 0.45
        normalized["unit"] = "kg"
    
    return normalized


def validateManifest(manifest):
    """
    Validates a cargo manifest and returns validation errors
    
    Args:
        manifest (dict): The cargo manifest object
        
    Returns:
        dict: An object containing missing/invalid properties, or empty if valid
    """
    errors = {}
    
    # Check containerId: must be a positive integer
    if "containerId" not in manifest:
        errors["containerId"] = "Missing"
    elif not isinstance(manifest["containerId"], int) or manifest["containerId"] <= 0:
        errors["containerId"] = "Invalid"
    
    # Check destination: must be a non-empty string after trimming
    if "destination" not in manifest:
        errors["destination"] = "Missing"
    elif not isinstance(manifest["destination"], str) or manifest["destination"].strip() == "":
        errors["destination"] = "Invalid"
    
    # Check weight: must be a positive number (not NaN)
    if "weight" not in manifest:
        errors["weight"] = "Missing"
    elif not isinstance(manifest["weight"], (int, float)) or manifest["weight"] != manifest["weight"] or manifest["weight"] <= 0:
        # manifest["weight"] != manifest["weight"] checks for NaN (NaN != NaN is True)
        errors["weight"] = "Invalid"
    
    # Check unit: must be "kg" or "lb"
    if "unit" not in manifest:
        errors["unit"] = "Missing"
    elif manifest["unit"] not in ["kg", "lb"]:
        errors["unit"] = "Invalid"
    
    # Check hazmat: must be a boolean
    if "hazmat" not in manifest:
        errors["hazmat"] = "Missing"
    elif not isinstance(manifest["hazmat"], bool):
        errors["hazmat"] = "Invalid"
    
    return errors


def processManifest(manifest):
    """
    Processes a cargo manifest, logging validation results
    
    Args:
        manifest (dict): The cargo manifest object
    """
    errors = validateManifest(manifest)
    is_valid = len(errors) == 0
    
    if is_valid:
        print(f"Validation success: {manifest['containerId']}")
        normalized = normalizeUnits(manifest)
        print(f"Total weight: {normalized['weight']} kg")
    else:
        print(f"Validation error: {manifest.get('containerId')}")
        print(errors)


# Example usage and testing
if __name__ == "__main__":
    # Test normalizeUnits
    test_manifest = {
        "containerId": 68,
        "destination": "Salinas",
        "weight": 101,
        "unit": "lb",
        "hazmat": True
    }
    
    print("Original:", test_manifest)
    normalized = normalizeUnits(test_manifest)
    print("Normalized:", normalized)
    print("Original after normalization (should be unchanged):", test_manifest)
    
    print("\n" + "="*60 + "\n")
    
    # Test validateManifest with valid manifest
    valid_manifest = {
        "containerId": 1,
        "destination": "Santa Cruz",
        "weight": 304,
        "unit": "kg",
        "hazmat": False
    }
    print("Validating valid manifest:", validateManifest(valid_manifest))
    
    # Test validateManifest with empty object
    print("Validating empty manifest:", validateManifest({}))
    
    # Test validateManifest with invalid containerId
    invalid_container = {
        "containerId": None,
        "destination": "Santa Cruz",
        "weight": 304,
        "unit": "kg",
        "hazmat": False
    }
    print("Validating invalid containerId:", validateManifest(invalid_container))
    
    # Test processManifest
    print("\n" + "="*60 + "\n")
    print("Processing valid manifest:")
    processManifest({
        "containerId": 55,
        "destination": "Carmel",
        "weight": 400,
        "unit": "lb",
        "hazmat": False
    })
    
    print("\nProcessing invalid manifest:")
    processManifest({
        "containerId": -88,
        "destination": "Soledad",
        "weight": float('nan')
    })

File: basic.java
// Cargo Manifest Validator - Java Implementation

import java.util.HashMap;
import java.util.Map;

public class CargoManifestValidator {
    
    /**
     * Represents a cargo manifest
     */
    public static class Manifest {
        public Integer containerId;
        public String destination;
        public Double weight;
        public String unit;
        public Boolean hazmat;
        
        public Manifest() {}
        
        public Manifest(Integer containerId, String destination, Double weight, String unit, Boolean hazmat) {
            this.containerId = containerId;
            this.destination = destination;
            this.weight = weight;
            this.unit = unit;
            this.hazmat = hazmat;
        }
    }
    
    /**
     * Normalizes the weight to kilograms without mutating the original manifest
     * @param manifest The cargo manifest object
     * @return A new manifest object with weight in kg
     */
    public static Map<String, Object> normalizeUnits(Map<String, Object> manifest) {
        // Create a copy of the manifest to avoid mutation
        Map<String, Object> normalized = new HashMap<>(manifest);
        
        // Convert pounds to kilograms if necessary
        if ("lb".equals(normalized.get("unit"))) {
            Double weight = (Double) normalized.get("weight");
            normalized.put("weight", weight * 0.45);
            normalized.put("unit", "kg");
        }
        
        return normalized;
    }
    
    /**
     * Validates a cargo manifest and returns validation errors
     * @param manifest The cargo manifest object
     * @return A map containing missing/invalid properties, or empty if valid
     */
    public static Map<String, String> validateManifest(Map<String, Object> manifest) {
        Map<String, String> errors = new HashMap<>();
        
        // Check containerId: must be a positive integer
        if (!manifest.containsKey("containerId")) {
            errors.put("containerId", "Missing");
        } else {
            Object containerId = manifest.get("containerId");
            if (!(containerId instanceof Integer) || (Integer) containerId <= 0) {
                errors.put("containerId", "Invalid");
            }
        }
        
        // Check destination: must be a non-empty string after trimming
        if (!manifest.containsKey("destination")) {
            errors.put("destination", "Missing");
        } else {
            Object destination = manifest.get("destination");
            if (!(destination instanceof String) || ((String) destination).trim().isEmpty()) {
                errors.put("destination", "Invalid");
            }
        }
        
        // Check weight: must be a positive number (not NaN)
        if (!manifest.containsKey("weight")) {
            errors.put("weight", "Missing");
        } else {
            Object weight = manifest.get("weight");
            if (!(weight instanceof Number) || 
                Double.isNaN(((Number) weight).doubleValue()) ||
                ((Number) weight).doubleValue() <= 0) {
                errors.put("weight", "Invalid");
            }
        }
        
        // Check unit: must be "kg" or "lb"
        if (!manifest.containsKey("unit")) {
            errors.put("unit", "Missing");
        } else {
            Object unit = manifest.get("unit");
            if (!"kg".equals(unit) && !"lb".equals(unit)) {
                errors.put("unit", "Invalid");
            }
        }
        
        // Check hazmat: must be a boolean
        if (!manifest.containsKey("hazmat")) {
            errors.put("hazmat", "Missing");
        } else {
            Object hazmat = manifest.get("hazmat");
            if (!(hazmat instanceof Boolean)) {
                errors.put("hazmat", "Invalid");
            }
        }
        
        return errors;
    }
    
    /**
     * Processes a cargo manifest, logging validation results
     * @param manifest The cargo manifest object
     */
    public static void processManifest(Map<String, Object> manifest) {
        Map<String, String> errors = validateManifest(manifest);
        boolean isValid = errors.isEmpty();
        
        if (isValid) {
            System.out.println("Validation success: " + manifest.get("containerId"));
            Map<String, Object> normalized = normalizeUnits(manifest);
            System.out.println("Total weight: " + normalized.get("weight") + " kg");
        } else {
            System.out.println("Validation error: " + manifest.get("containerId"));
            System.out.println(errors);
        }
    }
    
    // Main method for testing
    public static void main(String[] args) {
        // Test normalizeUnits
        Map<String, Object> testManifest = new HashMap<>();
        testManifest.put("containerId", 68);
        testManifest.put("destination", "Salinas");
        testManifest.put("weight", 101.0);
        testManifest.put("unit", "lb");
        testManifest.put("hazmat", true);
        
        System.out.println("Original: " + testManifest);
        Map<String, Object> normalized = normalizeUnits(testManifest);
        System.out.println("Normalized: " + normalized);
        System.out.println("Original after normalization (should be unchanged): " + testManifest);
        
        System.out.println("\n" + "=".repeat(60) + "\n");
        
        // Test validateManifest with valid manifest
        Map<String, Object> validManifest = new HashMap<>();
        validManifest.put("containerId", 1);
        validManifest.put("destination", "Santa Cruz");
        validManifest.put("weight", 304.0);
        validManifest.put("unit", "kg");
        validManifest.put("hazmat", false);
        System.out.println("Validating valid manifest: " + validateManifest(validManifest));
        
        // Test validateManifest with empty map
        System.out.println("Validating empty manifest: " + validateManifest(new HashMap<>()));
        
        // Test validateManifest with invalid containerId
        Map<String, Object> invalidContainer = new HashMap<>();
        invalidContainer.put("containerId", null);
        invalidContainer.put("destination", "Santa Cruz");
        invalidContainer.put("weight", 304.0);
        invalidContainer.put("unit", "kg");
        invalidContainer.put("hazmat", false);
        System.out.println("Validating invalid containerId: " + validateManifest(invalidContainer));
        
        // Test processManifest
        System.out.println("\n" + "=".repeat(60) + "\n");
        System.out.println("Processing valid manifest:");
        Map<String, Object> processTest = new HashMap<>();
        processTest.put("containerId", 55);
        processTest.put("destination", "Carmel");
        processTest.put("weight", 400.0);
        processTest.put("unit", "lb");
        processTest.put("hazmat", false);
        processManifest(processTest);
        
        System.out.println("\nProcessing invalid manifest:");
        Map<String, Object> invalidProcess = new HashMap<>();
        invalidProcess.put("containerId", -88);
        invalidProcess.put("destination", "Soledad");
        invalidProcess.put("weight", Double.NaN);
        processManifest(invalidProcess);
    }
}
